src/brain/indextune/compressed_index_config_util.cpp

```diff
@@ -0,0 +1,213 @@
+  container.GetTransactionManager()->CommitTransaction(txn);
+
+  return sql_stmt_list;
+}
+
+std::unique_ptr<boost::dynamic_bitset<>>
+CompressedIndexConfigUtil::GenerateBitSet(
+    const CompressedIndexConfigContainer &container,
+    const std::vector<std::shared_ptr<brain::HypotheticalIndexObject>>
+        &idx_objs) {
+  auto result = std::unique_ptr<boost::dynamic_bitset<>>(
+      new boost::dynamic_bitset<>(container.GetConfigurationCount()));
+
+  for (const auto &idx_obj : idx_objs) {
+    SetBit(container, *result, idx_obj);
+  }
+
+  return result;
+}
+
+void CompressedIndexConfigUtil::SetBit(
+    const CompressedIndexConfigContainer &container,
+    boost::dynamic_bitset<> &bitmap,
+    const std::shared_ptr<HypotheticalIndexObject> &idx_object) {
+  size_t offset = container.GetGlobalOffset(idx_object);
+  bitmap.set(offset);
+}
+
+void CompressedIndexConfigUtil::ConstructQueryConfigFeature(
+    const boost::dynamic_bitset<> &curr_config_set,
+    const boost::dynamic_bitset<> &add_candidate_set,
+    const boost::dynamic_bitset<> &drop_candidate_set,
+    vector_eig &query_config_vec) {
+  size_t num_configs = curr_config_set.size();
+  query_config_vec = vector_eig::Zero(2 * num_configs);
+  size_t offset_rec = 0;
+  size_t config_id_rec = add_candidate_set.find_first();
+  query_config_vec[offset_rec] = 1.0;
+  while (config_id_rec != boost::dynamic_bitset<>::npos) {
+    if (curr_config_set.test(config_id_rec)) {
+      query_config_vec[offset_rec + config_id_rec] = 1.0f;
+    } else {
+      query_config_vec[offset_rec + config_id_rec] = -1.0f;
+    }
+    config_id_rec = add_candidate_set.find_next(config_id_rec);
+  }
+  size_t offset_drop = num_configs;
+  size_t config_id_drop = drop_candidate_set.find_first();
+  query_config_vec[offset_drop] = 1.0;
+  while (config_id_drop != boost::dynamic_bitset<>::npos) {
+    if (curr_config_set.test(config_id_drop)) {
+      query_config_vec[offset_drop + config_id_drop] = 1.0f;
+    }
+    // else case shouldnt happen
+    config_id_drop = drop_candidate_set.find_next(config_id_drop);
+  }
+}
+
+void CompressedIndexConfigUtil::GetIgnoreTables(const std::string &db_name,
+                                             std::set<oid_t> &ori_table_oids) {
+  peloton::concurrency::TransactionManager *txn_manager =
+      &concurrency::TransactionManagerFactory::GetInstance();
+
+  auto txn = txn_manager->BeginTransaction();
+  const auto table_objs = catalog::Catalog::GetInstance()
+                              ->GetDatabaseObject(db_name, txn)
+                              ->GetTableObjects();
+
+  for (const auto &it : table_objs) {
+    ori_table_oids.insert(it.first);
+  }
+
+  txn_manager->CommitTransaction(txn);
+}
+
+void CompressedIndexConfigUtil::ToEigen(
+    const boost::dynamic_bitset<> &config_set, vector_eig &config_vec) {
+  // Note that the representation is reversed - but this should not affect
+  // anything
+  config_vec = vector_eig::Zero(config_set.size());
```

Comment:
I think for the indexes that are not in the current configuration, the feature mapping is 1 instead of -1? Refer to the 8th page of the paper.

Reply:
Yep - your right will fix this.